### PR TITLE
Update workshops to allow virtual emails, email content updates

### PR DESCRIPTION
**Pull request by tim-dot-org**
This PR updates the workshop email system to allow virtual workshops to send automated emails for: enrollment, 3-day reminders, 10-day reminders, and exit surveys. It updates the content of the emails to be appropriate for virtual workshops. It also makes some small changes to the workshop editing form and enrollment page to make the location field more user friendly.

Missing:
- Some links are just placeholders right now
- Some links still need to be updated to current year

## Links

- spec: [virtual workshop email spec](https://docs.google.com/document/d/1FiabZUosFLA7DsLBLeDFjZivY8C2MIRhPNPS-aKW8F8/edit)
- jira ticket: [virtual workshop email ticket](https://codedotorg.atlassian.net/browse/PLC-1152)
- intended email content (WIP): [virtual emails](https://docs.google.com/document/d/1Yg4czg3FH1LQXv1P4u4B8kkmhWJIKWSeoi8sH0G0jbo/edit#)

## Testing story

Unit tests updated to match new functionality.
Emails manually tested on local environment using rails mailer previews.

## Deployment strategy

*Unfinished PR - do not deploy*
Doesn't need to be deployed until mid-May, when the emails will actually be triggered.

## Follow-up work

Update summer workshop email content: [PLC-1164
](https://codedotorg.atlassian.net/browse/PLC-1164)

## PR Checklist:

- [x] Tests provide adequate coverage
- [X] Privacy and Security impacts have been assessed
- [x] Code is well-commented
- [X] New features are translatable or updates will not break translations
- [x] Relevant documentation has been added or updated
- [x] User impact is well-understood and desirable
- [X] Pull Request is labeled appropriately
- [X] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx [hunk 1]
```diff
@@ -983,19 +979,21 @@ export class WorkshopForm extends React.Component {
...
-                  placeholder="Enter a location"
-                  onChange={this.handleLocationChange}
-                  value={this.state.location_address || ''}
-                  readOnly={this.props.readOnly}
-                  style={this.getInputStyle()}
-                  className={'form-control'}
-                />
-              </FormGroup>
-            </Col>
+            {!this.state.virtual && (
```
Comment (bencodeorg) [inline apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx]: Is this an ask from the PL team? I'm wondering if workshop organizers like being able to put something obvious here that indicates that it's a virtual workshop, other than the join link?

File: apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx [hunk 2]
```diff
@@ -143,7 +152,9 @@ export default class WorkshopDetails extends React.Component {
           </div>
         </div>
         {this.sessionDates()}
-        {this.location()}
+        {this.props.hide_virtual_location && this.props.workshop.virtual
```
Comment (bencodeorg) [inline apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx]: Seems like the pattern here is encapsulate logic in the `location()` function?

File: apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx [hunk 3]
```diff
@@ -17,7 +17,9 @@ const styles = {
 export default class WorkshopDetails extends React.Component {
   static propTypes = {
     workshop: WorkshopPropType,
-    session_dates: PropTypes.arrayOf(PropTypes.string)
+    session_dates: PropTypes.arrayOf(PropTypes.string),
+    /* hides the location field in places we might not want to show it */
+    hide_virtual_location: PropTypes.bool
```
Comment (bencodeorg) [inline apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx]: Why add a separate prop here instead of using workshop.virtual?

File: dashboard/app/views/pd/workshop_mailer/_workshop_logistics.html.haml [hunk 4]
```diff
@@ -8,14 +8,21 @@
...
+  - unless @workshop.location_name.blank?
+    %tr
+      %td.heading
+        - if @workshop.virtual?
+          Link to join:
+        - else
+          Location:
+      %td
+        - if @workshop.virtual?
+          = link_to @workshop.location_name, @workshop.location_name, target: "_blank"
```
Comment (bencodeorg) [inline dashboard/app/views/pd/workshop_mailer/_workshop_logistics.html.haml]: I know we have some thoughts in the team about opening links in new tabs, would be nice to update this to align with those:

https://github.com/code-dot-org/code-dot-org/pull/37791

File: apps/test/unit/code-studio/pd/workshop_dashboard/workshopTest.jsx [hunk 5]
```diff
@@ -22,7 +22,7 @@ describe('Workshop', () => {
   it('renders', () => {
     const wrapper = shallow(
       <Workshop
-        params={{workshopId: fakeWorkshop.id}}
+        params={{workshopId: fakeWorkshop.id.toString()}}
```
Comment (bencodeorg) [inline apps/test/unit/code-studio/pd/workshop_dashboard/workshopTest.jsx:25]: For my understanding, why did you move this?

File: apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import '../workshopFactory';
+
+describe('WorkshopForm test', () => {
+  let fakeWorkshop;
+
+  beforeEach(() => {
+    fakeWorkshop = Factory.build('workshop');
+  });
+
+  afterEach(() => {});
```
Comment (bencodeorg) [inline apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js]: Is this necessary?

File: apps/test/unit/code-studio/pd/workshop_dashboard/workshopTest.jsx (hunk 5 above)
Comment (bencodeorg) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopTest.jsx:25]: I feel like I'm going to learn a lot about JS testing from you :)

File: apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        permission={new Permission()}
+        facilitatorCourses={[]}
+        workshop={fakeWorkshop}
+        onSaved={() => {}}
+        readOnly={false}
+      />,
+      {context: {router: {}}}
+    );
+
+    const locationGroup = wrapper.find('#location_name').closest('FormGroup');
```
Comment (bencodeorg) [inline apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js]: I've been encouraged to import the component itself and use that in Enzyme searches, but I don't really have a strong opinion on the matter...

File: apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
+import React from 'react';
+import {shallow} from 'enzyme';
+import {assert, expect} from 'chai';
+import {Factory} from 'rosie';
+import Permission from '@cdo/apps/code-studio/pd/workshop_dashboard/permission';
+import {WorkshopForm} from '@cdo/apps/code-studio/pd/workshop_dashboard/components/workshop_form';
+import '../workshopFactory';
+
+describe('WorkshopForm test', () => {
```
Comment (bencodeorg) [inline apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js:10]: We have no `WorkshopForm` tests?? Yikes! Thanks for adding!

**Review by bencodeorg — COMMENTED** [review]
This is great, and way more work than I was imagining for a first PR! Very impressive testing as well.

Have a few small questions for you -- feel a little bit weird about taking the "location name" attribute and using that for URLs, but if that's sort of happening organically (ie, partners have already co-opted it for that use) already then maybe it's ok.

File: apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx (hunk 3 above)
Comment (tim-dot-org) [reply apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx]: It seemed like this component could be reused elsewhere (it's not currently), and we wanted to specifically only hide the virtual address on the enrollment page (since the teacher isn't actually enrolled in the session yet)

File: apps/test/unit/code-studio/pd/workshop_dashboard/workshopTest.jsx (hunk 5 above)
Comment (tim-dot-org) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopTest.jsx:25]: The react props enforce that this field is actually a number on the "workshop" objects, it only needs to be a string for the workshopId fields, so when you pass the whole workshop, it should be a number by default it think

Comment (tim-dot-org) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopTest.jsx:25]: I learned a lot about testing when updating the tests 😅 I haven't used enzyme and chai before

File: apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js (hunk 7 above)
Comment (tim-dot-org) [reply apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js]: my reasoning is that its easier and reads better to grab by id instead of `.find(FormGroup).childAt(3)`, plus it'd break if the fields got reordered

File: apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js (hunk 6 above)
Comment (tim-dot-org) [reply apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js]: nope, left over from copy, dropping it

File: dashboard/app/views/pd/workshop_mailer/_workshop_logistics.html.haml (hunk 4 above)
Comment (tim-dot-org) [reply dashboard/app/views/pd/workshop_mailer/_workshop_logistics.html.haml]: oh cool, i learned something! I'll update the other links too

File: apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx (hunk 1 above)
Comment (tim-dot-org) [reply apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx]: Yeah, from Tess:
`This is our regional partners using a field in a way it wasnt designed.  Theres about 200 already scheduled workshops so we probably won't have success if we need partners to go through and change this before summer workshops.   We should talk about the right way to do this before Annual Year Workshops start getting scheduled, but for now we are in a bit of limbo.
`

Comment (tim-dot-org) [reply apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx]: Tess and I talked about a couple of options, and decided to use the thing the partners were doing already for now, but I think it'd make more sense as a separate field in the long term!

File: apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js (hunk 7 above)
Comment (tim-dot-org) [reply apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js]: oh but yeah, i can import the component for FormGroup!

File: apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx (hunk 1 above)
Comment (bencodeorg) [reply apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx]: That makes sense -- I'm curious if we're worried about how this will read for virtual workshops that have already been created, with location names that aren't links? Like this one?

https://studio.code.org/pd/workshop_dashboard/workshops/10273

Comment (tim-dot-org) [reply apps/src/code-studio/pd/workshop_dashboard/components/workshop_form.jsx]: oh yeah, good point, it looks like they're sending emails themselves, and are probably just putting the link to join there. 
we could parse the location_name and check for a url, which would add some complexity but make it more durable for any case, happy to do that
@tess323 any thoughts?

File: apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js (hunk 7 above)
Comment (bencodeorg) [reply apps/test/unit/code-studio/pd/workshop_dashboard/components/workshopFormTest.js]: Interesting -- I do think we do a lot of index type searches, and this does seem more robust! I guess if you can use the component in the that would be nice but not a big deal.

File: apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx (hunk 3 above)
Comment (bencodeorg) [reply apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx]: I guess I don't totally understand -- we're using `this.props.workshop.virtual` to decide what description to show ("Link to join" or "Location"), then the new prop for whether to actually show the location?

File: apps/test/unit/code-studio/pd/workshop_dashboard/workshopTest.jsx (hunk 5 above)
Comment (bencodeorg) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopTest.jsx:25]: haha you could have fooled me!

Comment (bencodeorg) [reply apps/test/unit/code-studio/pd/workshop_dashboard/workshopTest.jsx:25]: Ooh I see what you're saying about workshop ID, interesting -- no idea why we use integers in one place and strings in another, but that's for another day. Nice cleanup!

**Comment by bencodeorg** [comment]
Ooh and one other general comment before I forget -- I think the PR description looks great, but I think we could use a more descriptive title here :)

File: apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx (hunk 3 above)
Comment (tim-dot-org) [reply apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx]: Right, for the hypothetical use case where this component gets rendered somewhere else and we want to show the location link. It's maybe a bit too complicated for something that isn't actually needed, haha, I can take it out

Comment (bencodeorg) [reply apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx]: Yeah I may be being thick but I'd prefer to take it out if we're not using it :)

File: dashboard/test/controllers/api/v1/pd/workshops_controller_test.rb [hunk 9]
```diff
@@ -677,15 +679,16 @@ class Api::V1::Pd::WorkshopsControllerTest < ::ActionController::TestCase
     assert workshop.virtual?
   end
 
-  test 'cannot update a workshop to be virtual without suppressed email' do
+  test 'can update a workshop to be virtual without suppressed email' do
```
Comment (bencodeorg) [inline dashboard/test/controllers/api/v1/pd/workshops_controller_test.rb:684]: I liked how in another test you updated like this that you noted this test exists because it's a change to a previous behavior, could you add something like that here?

**Review by bencodeorg — APPROVED** [review]
Looks great! 🎉

File: apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx [hunk 10]
```diff
@@ -53,15 +53,22 @@ export default class WorkshopDetails extends React.Component {
   }
 
   location() {
-    return (
+    return this.props.workshop.virtual ? null : (
```
Comment (bencodeorg) [inline apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx]: Ack I'm still not understanding this! We have a check here that returns nothing if the workshop is virtual, but a check in the returned HTML when virtual = false for whether to show "Link to join" or "Location"?

Comment (tim-dot-org) [reply apps/src/code-studio/pd/workshop_enrollment/workshop_details.jsx]: oh, good catch, i forgot to change that too

File: dashboard/test/controllers/api/v1/pd/workshops_controller_test.rb (hunk 9 above)
Comment (tim-dot-org) [reply dashboard/test/controllers/api/v1/pd/workshops_controller_test.rb:684]: yep!

**Comment by tess323** [comment]
Hey team - sorry for the late change.  Spoke with Tawny and outreach.  Can we remove the location field all together from the virtual emails.  Partners aren't using it reliably in the way we want.  We will make sure they are and then add this field for the annual year workshop.  Thank you both for flagging this and apologies again for the late change.

**Comment by tim-dot-org** [comment]
> 
> 
> Hey team - sorry for the late change. Spoke with Tawny and outreach. Can we remove the location field all together from the virtual emails. Partners aren't using it reliably in the way we want. We will make sure they are and then add this field for the annual year workshop. Thank you both for flagging this and apologies again for the late change.

Sounds good! I'll revert the changes to the form as well, and leave it as just location and address.

**Comment by bencodeorg** [comment]
@tess323 just to clarify around what we want to show on workshop enrollment pages for virtual workshops -- this PR I think will remove the "Location" row entirely from the enrollment page for virtual workshops. Is that what we want (even if partners have used it to mark that a workshop is virtual? Maybe we could include a line on the enroll page that says `Location: Virtual` for any virtual workshop?

https://studio.code.org/pd/workshops/9859/enroll

**Comment by tess323** [comment]
I like that solution - lets go with Location: Virtual  for any virtual workshop.  Thank you all for weeding through this with me.

**Comment by tim-dot-org** [comment]
rebased on staging, just added the last commit to update the pre-reading content and add the order form link

**Review by bencodeorg — APPROVED** [review]
Sorry I didn't see the request, LGTM!

**Comment by tim-dot-org** [comment]
> 
> 
> Sorry I didn't see the request, LGTM!

no worries! not going to merge until i have the teacher export anyways
